File: infrastructure/runtime/src/portability/export.ts
// Agent export — produce a portable AgentFile from a running nous
//
// Captures: config, workspace files, session history, working state,
// agent notes, distillation priming, and optionally memory vectors + graph.
//
// Design: Spec 21 Phase 1 (Agent Portability)

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { extname, join, relative } from "node:path";
import { createLogger } from "../koina/logger.js";
import type { DistillationPriming, Session, SessionStore, WorkingState } from "../mneme/store.js";
import { paths } from "../taxis/paths.js";

const log = createLogger("portability:export");

// --- AgentFile Schema ---

export interface AgentFile {
  version: 1;
  exportedAt: string;
  generator: string;
  nous: {
    id: string;
    name: string | null;
    model: string | null;
    config: Record<string, unknown>;
  };
  workspace: {
    files: Record<string, string>;    // path → content (text files only)
    binaryFiles: string[];            // paths of binary files (not included)
  };
  sessions: ExportedSession[];
  memory?: {
    vectors?: ExportedVector[];
    graph?: {
      nodes: ExportedGraphNode[];
      edges: ExportedGraphEdge[];
    };
  };
}

export interface ExportedSession {
  id: string;
  sessionKey: string;
  status: Session["status"];
  sessionType: Session["sessionType"];
  messageCount: number;
  tokenCountEstimate: number;
  distillationCount: number;
  createdAt: string;
  updatedAt: string;
  workingState: WorkingState | null;
  distillationPriming: DistillationPriming | null;
  notes: ExportedNote[];
  messages: ExportedMessage[];
}

export interface ExportedMessage {
  role: string;
  content: string;
  seq: number;
  tokenEstimate: number;
  isDistilled: boolean;
  createdAt: string;
}

export interface ExportedNote {
  category: string;
  content: string;
  createdAt: string;
}

export interface ExportedVector {
  id: string;
  text: string;
  metadata: Record<string, unknown>;
  // Embeddings omitted — can be regenerated on import
}

export interface ExportedGraphNode {
  name: string;
  labels: string[];
  properties: Record<string, unknown>;
}

export interface ExportedGraphEdge {
  source: string;
  target: string;
  relType: string;
}

// --- Export Options ---

export interface ExportOptions {
  /** Include memory vectors from Qdrant */
  includeMemory?: boolean;
  /** Include knowledge graph from Neo4j */
  includeGraph?: boolean;
  /** Max messages per session (0 = all). Default: 500 */
  maxMessagesPerSession?: number;
  /** Include archived sessions. Default: false */
  includeArchived?: boolean;
  /** Memory sidecar URL for vector/graph export */
  sidecarUrl?: string;
}

const DEFAULT_OPTIONS: Required<ExportOptions> = {
  includeMemory: false,
  includeGraph: false,
  maxMessagesPerSession: 500,
  includeArchived: false,
  sidecarUrl: "http://localhost:8230",
};

// --- Text file detection ---

const TEXT_EXTENSIONS = new Set([
  ".md", ".txt", ".yaml", ".yml", ".json", ".ts", ".js", ".py",
  ".sh", ".bash", ".zsh", ".css", ".html", ".svelte", ".toml",
  ".ini", ".cfg", ".conf", ".env", ".log", ".csv", ".xml",
  ".gitignore", ".editorconfig", ".prettierrc",
]);

const BINARY_EXTENSIONS = new Set([
  ".png", ".jpg", ".jpeg", ".gif", ".ico", ".svg", ".webp",
  ".woff", ".woff2", ".ttf", ".eot",
  ".zip", ".tar", ".gz", ".bz2", ".xz",
  ".pdf", ".doc", ".docx", ".xlsx",
  ".db", ".sqlite", ".sqlite3",
  ".wasm", ".so", ".dylib",
]);

function isTextFile(filename: string): boolean {
  const ext = extname(filename).toLowerCase();
  if (TEXT_EXTENSIONS.has(ext)) return true;
  if (BINARY_EXTENSIONS.has(ext)) return false;
  // No extension or unknown — check if no extension (likely text: Makefile, Dockerfile, etc.)
  if (!ext) return true;
  return false; // Default to binary for unknown extensions
}

// --- Workspace scanning ---

const IGNORE_DIRS = new Set(["node_modules", ".git", "__pycache__", ".cache", "dist"]);
const MAX_FILE_SIZE = 1024 * 1024; // 1MB — skip files larger than this

function scanWorkspace(
  workspacePath: string,
): { files: Record<string, string>; binaryFiles: string[] } {
  const files: Record<string, string> = {};
  const binaryFiles: string[] = [];

  function walk(dir: string): void {
    let entries: string[];
    try {
      entries = readdirSync(dir);
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.startsWith(".") && entry !== ".env") continue;
      if (IGNORE_DIRS.has(entry)) continue;

      const fullPath = join(dir, entry);
      let stat;
      try {
        stat = statSync(fullPath);
      } catch {
        continue;
      }

      if (stat.isDirectory()) {
        walk(fullPath);
        continue;
      }

      if (!stat.isFile()) continue;

      const relPath = relative(workspacePath, fullPath);

      if (stat.size > MAX_FILE_SIZE) {
        binaryFiles.push(relPath);
        continue;
      }

      if (isTextFile(entry)) {
        try {
          files[relPath] = readFileSync(fullPath, "utf-8");
        } catch {
          binaryFiles.push(relPath);
        }
      } else {
        binaryFiles.push(relPath);
      }
    }
  }

  walk(workspacePath);
  return { files, binaryFiles };
}

// --- Session export ---

function exportSession(
  store: SessionStore,
  session: Session,
  _nousId: string,
  maxMessages: number,
): ExportedSession {
  // Get messages — most recent N if limited
  const limit = maxMessages > 0 ? maxMessages : 0;
  const messages = store.getHistory(session.id, limit > 0 ? { limit } : {});

  // Get notes for this session
  const notes = store.getNotes(session.id, { limit: 100 });

  return {
    id: session.id,
    sessionKey: session.sessionKey,
    status: session.status,
    sessionType: session.sessionType,
    messageCount: session.messageCount,
    tokenCountEstimate: session.tokenCountEstimate,
    distillationCount: session.distillationCount,
    createdAt: session.createdAt,
    updatedAt: session.updatedAt,
    workingState: session.workingState,
    distillationPriming: session.distillationPriming,
    notes: notes.map((n) => ({
      category: n.category,
      content: n.content,
      createdAt: n.createdAt,
    })),
    messages: messages.map((m) => ({
      role: m.role,
      content: m.content,
      seq: m.seq,
      tokenEstimate: m.tokenEstimate,
      isDistilled: m.isDistilled,
      createdAt: m.createdAt,
    })),
  };
}

// --- Memory export (via sidecar HTTP) ---

async function exportMemoryVectors(
  sidecarUrl: string,
  nousId: string,
): Promise<ExportedVector[]> {
  try {
    const response = await fetch(
      `${sidecarUrl}/memories?agent_id=${encodeURIComponent(nousId)}&limit=10000`,
    );
    if (!response.ok) {
      log.warn(`Memory export failed: ${response.status} ${response.statusText}`);
      return [];
    }
    const data = (await response.json()) as { ok: boolean; memories?: Array<Record<string, unknown>> };
    if (!data.ok || !data.memories) return [];

    return data.memories.map((m) => ({
      id: String(m["id"] ?? ""),
      text: String(m["memory"] ?? m["text"] ?? ""),
      metadata: (m["metadata"] as Record<string, unknown>) ?? {},
    }));
  } catch (err) {
    log.warn(`Memory sidecar unreachable: ${err instanceof Error ? err.message : err}`);
    return [];
  }
}

async function exportGraph(
  sidecarUrl: string,
): Promise<{ nodes: ExportedGraphNode[]; edges: ExportedGraphEdge[] } | null> {
  try {
    const response = await fetch(`${sidecarUrl}/graph/export?mode=all`);
    if (!response.ok) {
      log.warn(`Graph export failed: ${response.status} ${response.statusText}`);
      return null;
    }
    const data = (await response.json()) as {
      ok: boolean;
      nodes?: Array<Record<string, unknown>>;
      edges?: Array<Record<string, unknown>>;
    };
    if (!data.ok) return null;

    const nodes: ExportedGraphNode[] = (data.nodes ?? []).map((n) => ({
      name: String(n["id"] ?? n["name"] ?? ""),
      labels: (n["labels"] as string[]) ?? [],
      properties: {
        pagerank: n["pagerank"],
        community: n["community"],
      },
    }));

    const edges: ExportedGraphEdge[] = (data.edges ?? []).map((e) => ({
      source: String(e["source"] ?? ""),
      target: String(e["target"] ?? ""),
      relType: String(e["rel_type"] ?? e["relType"] ?? "RELATED_TO"),
    }));

    return { nodes, edges };
  } catch (err) {
    log.warn(`Graph export failed: ${err instanceof Error ? err.message : err}`);
    return null;
  }
}

// --- Main export function ---

export async function exportAgent(
  nousId: string,
  nousConfig: Record<string, unknown>,
  store: SessionStore,
  opts?: ExportOptions,
): Promise<AgentFile> {
  const options = { ...DEFAULT_OPTIONS, ...opts };

  log.info(`Exporting agent ${nousId}...`);

  // 1. Workspace files
  const workspacePath = paths.nousDir(nousId);
  let workspace = { files: {} as Record<string, string>, binaryFiles: [] as string[] };
  if (existsSync(workspacePath)) {
    workspace = scanWorkspace(workspacePath);
    log.info(`Workspace: ${Object.keys(workspace.files).length} text files, ${workspace.binaryFiles.length} binary files`);
  } else {
    log.warn(`Workspace not found: ${workspacePath}`);
  }

  // 2. Sessions
  const allSessions = store.listSessions(nousId);
  const sessions = options.includeArchived
    ? allSessions
    : allSessions.filter((s) => s.status !== "archived");

  const exportedSessions = sessions.map((s) =>
    exportSession(store, s, nousId, options.maxMessagesPerSession),
  );
  log.info(`Sessions: ${exportedSessions.length} (${sessions.length} total, ${allSessions.length - sessions.length} archived skipped)`);

  // 3. Memory (optional)
  let memory: AgentFile["memory"];
  if (options.includeMemory || options.includeGraph) {
    memory = {};

    if (options.includeMemory) {
      memory.vectors = await exportMemoryVectors(options.sidecarUrl, nousId);
      log.info(`Memory vectors: ${memory.vectors.length}`);
    }

    if (options.includeGraph) {
      const graph = await exportGraph(options.sidecarUrl);
      if (graph) {
        memory.graph = graph;
        log.info(`Graph: ${graph.nodes.length} nodes, ${graph.edges.length} edges`);
      }
    }
  }

  // 4. Build AgentFile
  const agentFile: AgentFile = {
    version: 1,
    exportedAt: new Date().toISOString(),
    generator: `aletheia-export/1.0`,
    nous: {
      id: nousId,
      name: (nousConfig["name"] as string) ?? null,
      model: (nousConfig["model"] as string) ?? null,
      config: nousConfig,
    },
    workspace,
    sessions: exportedSessions,
  };

  if (memory) {
    agentFile.memory = memory;
  }

  const jsonSize = JSON.stringify(agentFile).length;
  log.info(`Export complete: ${(jsonSize / 1024 / 1024).toFixed(1)}MB`);

  return agentFile;
}

// --- CLI entry point helper ---

export function agentFileToJson(agentFile: AgentFile, pretty = true): string {
  return JSON.stringify(agentFile, null, pretty ? 2 : undefined);
}

File: crates/mneme/src/export.rs
//! Agent export: build an `AgentFile` from session store and workspace.
#![cfg_attr(
    any(feature = "mneme-engine", test),
    expect(
        clippy::indexing_slicing,
        reason = "knowledge engine: ported codebase with numeric casts and direct indexing throughout"
    )
)]

use std::collections::HashMap;
use std::path::Path;

use snafu::ResultExt;
use tracing::{info, instrument, warn};

use crate::error::{self, Result};
use crate::portability::{
    AGENT_FILE_VERSION, AgentFile, ExportedMessage, ExportedNote, ExportedSession, NousInfo,
    WorkspaceData,
};
use crate::store::SessionStore;
use crate::types::SessionStatus;

/// Maximum file size to include in export (10 MB).
const MAX_FILE_SIZE: u64 = 10 * 1024 * 1024;

/// Byte scan window for binary detection.
const BINARY_PROBE_SIZE: usize = 8192;

/// Directories to skip during workspace scan.
const IGNORE_DIRS: &[&str] = &[
    ".git",
    "node_modules",
    "target",
    "__pycache__",
    ".cache",
    "dist",
];

/// Options controlling what gets exported.
#[derive(Debug, Clone)]
pub struct ExportOptions {
    /// Maximum messages per session (0 = all).
    pub max_messages_per_session: usize,
    /// Include archived/distilled sessions.
    pub include_archived: bool,
}

impl Default for ExportOptions {
    fn default() -> Self {
        Self {
            max_messages_per_session: 500,
            include_archived: false,
        }
    }
}

/// Export an agent to a portable `AgentFile`.
///
/// The caller resolves paths and config from Oikos/taxis, then passes
/// simple types here: mneme never touches taxis.
///
/// # Errors
///
/// Returns an error if session store queries or workspace I/O fails.
#[instrument(skip(store))]
pub fn export_agent(
    nous_id: &str,
    agent_name: Option<&str>,
    agent_model: Option<&str>,
    agent_config: serde_json::Value,
    store: &SessionStore,
    workspace_path: &Path,
    opts: &ExportOptions,
) -> Result<AgentFile> {
    let workspace = scan_workspace(workspace_path)?;

    let all_sessions = store.list_sessions(Some(nous_id))?;
    let filtered: Vec<_> = if opts.include_archived {
        all_sessions
    } else {
        all_sessions
            .into_iter()
            .filter(|s| s.status == SessionStatus::Active)
            .collect()
    };

    let mut sessions = Vec::with_capacity(filtered.len());
    for session in &filtered {
        sessions.push(export_session(store, session, opts)?);
    }

    let exported_at = jiff::Timestamp::now()
        .strftime("%Y-%m-%dT%H:%M:%SZ")
        .to_string();
    let generator = format!("aletheia-rust/{}", env!("CARGO_PKG_VERSION"));

    info!(
        nous_id,
        sessions = sessions.len(),
        workspace_files = workspace.files.len(),
        binary_files = workspace.binary_files.len(),
        "agent exported"
    );

    Ok(AgentFile {
        version: AGENT_FILE_VERSION,
        exported_at,
        generator,
        nous: NousInfo {
            id: nous_id.to_owned(),
            name: agent_name.map(String::from),
            model: agent_model.map(String::from),
            config: agent_config,
        },
        workspace,
        sessions,
        memory: None,
        knowledge: None,
    })
}

/// Build a `KnowledgeExport` from the knowledge store.
///
/// Queries all facts, entities, and relationships for the given nous.
/// Returns `None` if the store is empty or the query fails.
#[cfg(feature = "mneme-engine")]
#[instrument(skip(store))]
pub fn export_knowledge(
    nous_id: &str,
    store: &crate::knowledge_store::KnowledgeStore,
) -> Option<crate::portability::KnowledgeExport> {
    let facts = store
        .query_facts(nous_id, "9999-01-01T00:00:00Z", 100_000)
        .ok()
        .unwrap_or_default();

    let entities = query_all_entities(store).unwrap_or_default();

    let relationships = query_all_relationships(store).unwrap_or_default();

    if facts.is_empty() && entities.is_empty() && relationships.is_empty() {
        return None;
    }

    info!(
        nous_id,
        facts = facts.len(),
        entities = entities.len(),
        relationships = relationships.len(),
        "knowledge exported"
    );

    Some(crate::portability::KnowledgeExport {
        facts,
        entities,
        relationships,
    })
}

/// Query all entities from the knowledge store.
#[cfg(feature = "mneme-engine")]
fn query_all_entities(
    store: &crate::knowledge_store::KnowledgeStore,
) -> Result<Vec<crate::knowledge::Entity>> {
    use std::collections::BTreeMap;

    let script = r"?[id, name, entity_type, aliases, created_at, updated_at] := *entities{id, name, entity_type, aliases, created_at, updated_at}";
    let rows = store.run_query(script, BTreeMap::new())?;

    let mut entities = Vec::new();
    for row in &rows.rows {
        if row.len() < 6 {
            continue;
        }
        let id = crate::id::EntityId::new_unchecked(row[0].get_str().unwrap_or_default());
        let name = row[1].get_str().unwrap_or_default().to_owned();
        let entity_type = row[2].get_str().unwrap_or_default().to_owned();
        let aliases_str = row[3].get_str().unwrap_or_default();
        let aliases = if aliases_str.is_empty() {
            vec![]
        } else {
            aliases_str
                .split(',')
                .map(|s| s.trim().to_owned())
                .collect()
        };
        let created_at = crate::knowledge::parse_timestamp(row[4].get_str().unwrap_or_default())
            .unwrap_or_else(jiff::Timestamp::now);
        let updated_at = crate::knowledge::parse_timestamp(row[5].get_str().unwrap_or_default())
            .unwrap_or_else(jiff::Timestamp::now);

        entities.push(crate::knowledge::Entity {
            id,
            name,
            entity_type,
            aliases,
            created_at,
            updated_at,
        });
    }

    Ok(entities)
}

/// Query all relationships from the knowledge store.
#[cfg(feature = "mneme-engine")]
fn query_all_relationships(
    store: &crate::knowledge_store::KnowledgeStore,
) -> Result<Vec<crate::knowledge::Relationship>> {
    use std::collections::BTreeMap;

    let script = r"?[src, dst, relation, weight, created_at] := *relationships{src, dst, relation, weight, created_at}";
    let rows = store.run_query(script, BTreeMap::new())?;

    let mut relationships = Vec::new();
    for row in &rows.rows {
        if row.len() < 5 {
            continue;
        }
        let src = crate::id::EntityId::new_unchecked(row[0].get_str().unwrap_or_default());
        let dst = crate::id::EntityId::new_unchecked(row[1].get_str().unwrap_or_default());
        let relation = row[2].get_str().unwrap_or_default().to_owned();
        let weight = row[3].get_float().unwrap_or(0.0);
        let created_at = crate::knowledge::parse_timestamp(row[4].get_str().unwrap_or_default())
            .unwrap_or_else(jiff::Timestamp::now);

        relationships.push(crate::knowledge::Relationship {
            src,
            dst,
            relation,
            weight,
            created_at,
        });
    }

    Ok(relationships)
}

/// Scan a workspace directory, classifying files as text or binary.
fn scan_workspace(workspace_path: &Path) -> Result<WorkspaceData> {
    let mut files = HashMap::new();
    let mut binary_files = Vec::new();

    if !workspace_path.exists() {
        warn!(path = %workspace_path.display(), "workspace not found, exporting empty");
        return Ok(WorkspaceData {
            files,
            binary_files,
        });
    }

    walk_directory(
        workspace_path,
        workspace_path,
        &mut files,
        &mut binary_files,
    )?;

    Ok(WorkspaceData {
        files,
        binary_files,
    })
}

/// Recursive directory walk collecting text and binary file paths.
fn walk_directory(
    root: &Path,
    current: &Path,
    files: &mut HashMap<String, String>,
    binary_files: &mut Vec<String>,
) -> Result<()> {
    let entries = match std::fs::read_dir(current) {
        Ok(entries) => entries,
        Err(e) => {
            warn!(path = %current.display(), error = %e, "cannot read directory");
            return Ok(());
        }
    };

    for entry in entries {
        let entry = entry.context(error::IoSnafu {
            path: current.to_path_buf(),
        })?;
        let path = entry.path();
        let file_type = entry
            .file_type()
            .context(error::IoSnafu { path: path.clone() })?;

        if file_type.is_dir() {
            let dir_name = entry.file_name();
            let name = dir_name.to_string_lossy();
            if IGNORE_DIRS.contains(&name.as_ref()) {
                continue;
            }
            walk_directory(root, &path, files, binary_files)?;
            continue;
        }

        if !file_type.is_file() {
            continue;
        }

        let rel_path = path
            .strip_prefix(root)
            .unwrap_or(&path)
            .to_string_lossy()
            .replace('\\', "/");

        let metadata = match std::fs::metadata(&path) {
            Ok(m) => m,
            Err(e) => {
                warn!(path = %path.display(), error = %e, "cannot stat file");
                binary_files.push(rel_path);
                continue;
            }
        };

        if metadata.len() > MAX_FILE_SIZE {
            binary_files.push(rel_path);
            continue;
        }

        if is_binary_path(&path) || is_binary_content(&path) {
            binary_files.push(rel_path);
        } else {
            match std::fs::read_to_string(&path) {
                Ok(content) => {
                    files.insert(rel_path, content);
                }
                Err(_) => {
                    binary_files.push(rel_path);
                }
            }
        }
    }

    Ok(())
}

/// Check if a file path has a known binary extension.
fn is_binary_path(path: &Path) -> bool {
    const BINARY_EXTENSIONS: &[&str] = &[
        "png", "jpg", "jpeg", "gif", "ico", "svg", "webp", "woff", "woff2", "ttf", "eot", "zip",
        "tar", "gz", "bz2", "xz", "pdf", "doc", "docx", "xlsx", "db", "sqlite", "sqlite3", "wasm",
        "so", "dylib", "exe", "dll", "o", "a",
    ];

    path.extension()
        .and_then(|e| e.to_str())
        .is_some_and(|ext| BINARY_EXTENSIONS.contains(&ext.to_lowercase().as_str()))
}

/// Probe file content for null bytes indicating binary data.
fn is_binary_content(path: &Path) -> bool {
    use std::io::Read;

    #[expect(
        clippy::disallowed_methods,
        reason = "mneme filesystem operations access the embedded DB or model files; synchronous I/O is required in these contexts"
    )]
    let Ok(file) = std::fs::File::open(path) else {
        return true;
    };

    let mut buf = vec![0u8; BINARY_PROBE_SIZE];
    #[expect(
        clippy::as_conversions,
        reason = "usize→u64: BINARY_PROBE_SIZE is a small constant, fits in u64"
    )]
    let Ok(n) = file.take(BINARY_PROBE_SIZE as u64).read(&mut buf) else {
        return true;
    };

    #[expect(
        clippy::indexing_slicing,
        reason = "n comes from read() which guarantees n <= buf.len()"
    )]
    buf[..n].contains(&0)
}

/// Export a single session with all messages and notes.
fn export_session(
    store: &SessionStore,
    session: &crate::types::Session,
    opts: &ExportOptions,
) -> Result<ExportedSession> {
    let messages = get_all_messages(store, &session.id, opts.max_messages_per_session)?;
    let notes = store.get_notes(&session.id)?;

    let (working_state, distillation_priming) = get_session_json_fields(store, &session.id)?;

    Ok(ExportedSession {
        id: session.id.clone(),
        session_key: session.session_key.clone(),
        status: session.status.as_str().to_owned(),
        session_type: session.session_type.as_str().to_owned(),
        message_count: session.metrics.message_count,
        token_count_estimate: session.metrics.token_count_estimate,
        distillation_count: session.metrics.distillation_count,
        created_at: session.created_at.clone(),
        updated_at: session.updated_at.clone(),
        working_state,
        distillation_priming,
        notes: notes
            .into_iter()
            .map(|n| ExportedNote {
                category: n.category,
                content: n.content,
                created_at: n.created_at,
            })
            .collect(),
        messages,
    })
}

/// Get ALL messages for a session (including distilled) via raw SQL.
fn get_all_messages(
    store: &SessionStore,
    session_id: &str,
    max: usize,
) -> Result<Vec<ExportedMessage>> {
    let conn = store.conn();
    let sql = if max > 0 {
        "SELECT seq, role, content, token_estimate, is_distilled, created_at \
         FROM messages WHERE session_id = ?1 ORDER BY seq ASC LIMIT ?2"
    } else {
        "SELECT seq, role, content, token_estimate, is_distilled, created_at \
         FROM messages WHERE session_id = ?1 ORDER BY seq ASC"
    };

    let mut stmt = conn.prepare_cached(sql).context(error::DatabaseSnafu)?;

    let rows = if max > 0 {
        stmt.query_map(
            rusqlite::params![session_id, i64::try_from(max).unwrap_or(i64::MAX)],
            map_exported_message,
        )
        .context(error::DatabaseSnafu)?
    } else {
        stmt.query_map(rusqlite::params![session_id], map_exported_message)
            .context(error::DatabaseSnafu)?
    };

    let mut messages = Vec::new();
    for row in rows {
        messages.push(row.context(error::DatabaseSnafu)?);
    }
    Ok(messages)
}

/// Map a row to an `ExportedMessage`.
fn map_exported_message(row: &rusqlite::Row<'_>) -> rusqlite::Result<ExportedMessage> {
    let distilled: i64 = row.get(4)?;
    Ok(ExportedMessage {
        seq: row.get(0)?,
        role: row.get(1)?,
        content: row.get(2)?,
        token_estimate: row.get(3)?,
        is_distilled: distilled != 0,
        created_at: row.get(5)?,
    })
}

/// Read `working_state` and `distillation_priming` TEXT columns as JSON.
fn get_session_json_fields(
    store: &SessionStore,
    session_id: &str,
) -> Result<(Option<serde_json::Value>, Option<serde_json::Value>)> {
    let conn = store.conn();
    let mut stmt = conn
        .prepare_cached("SELECT working_state, distillation_priming FROM sessions WHERE id = ?1")
        .context(error::DatabaseSnafu)?;

    let result = stmt
        .query_row([session_id], |row| {
            let ws: Option<String> = row.get(0)?;
            let dp: Option<String> = row.get(1)?;
            Ok((ws, dp))
        })
        .context(error::DatabaseSnafu)?;

    let working_state = result.0.and_then(|s| serde_json::from_str(&s).ok());
    let distillation_priming = result.1.and_then(|s| serde_json::from_str(&s).ok());

    Ok((working_state, distillation_priming))
}

#[cfg(test)]
#[path = "export_tests.rs"]
mod tests;
